Made progress dots scroll horizontally underneath left and right arrows when they are too many

MAKOSCAFEE opened the pull request:
When progress dots are too many then I had to slice the array and show only certain range but the range keeps updating based on where  the current dot is. 
I changed from using ```$index``` to ```id```  because when using  ```$index```  its hard to keep track of sliced dots Eg. if sliced portion is from 6-10 but 6 will be at index zero and current page will be 0 instead of 6.
But i have made ```id``` start from zero too like ```$index``` so that it does not break any thing. 
I have chosen 25 to be maximum number of Dots  shown before they start to scroll horizontally underneath arrows @seanlip what do you think?.

seanlip commented:
I don't know if I like having two simultaneous indexing systems -- it could get confusing. How about adding a style like "display: none;" to the dots that are "out of range"? That way, all the dots will still be in the ng-repeat and you don't have to do lots of additional slicing.

Re number of dots, I think 25 is probably fine. Can you send a screenshot?

----

Also, I had a quick chat with @amitdeutsch about how exactly this should work from a design perspective. Here's what we came up with (note that you don't need to follow the arrow type in this screenshot):

![scrolling-dots](https://cloud.githubusercontent.com/assets/10575562/10006668/ff26d4ea-6073-11e5-8919-93bbd2a02ea4.png)

Note that this is for the case when there are currently 26 cards. The arrows at the end should be disabled somehow if clicking them won't do anything. Also, the dots 'scroll' only when the arrows are clicked and the currently active dot is next to the arrow that's being clicked. It would be nice if there was some animation there, but it's also fine to add it in a separate future pass if this is too tricky.

Also, the leftmost/rightmost dot should "fade out" (where relevant) in order to indicate that there's more stuff on either side. Probably best to use a left-to-right CSS gradient as described [here](http://www.w3schools.com/css/css3_gradients.asp).

Does this make sense? If you have any questions on the expected behaviour, please let us know!

MAKOSCAFEE commented:
Thanks @seanlip  for the feedback!
I have reverted back to using ```index``` instead of ```id``` . Dots is an array of empty objects, so depending on the maximum number of dots to be shown and the ```currentDotindex``` , range of objects(member of an array) to be assigned cssClassNames value is updated. These cssClassNames do two jobs
+  Either show or hide(when maximum number is exceeded)
+  animate the last  dot before hiding of other dots takes place.

I have added two types of "fade out" effects using css linear gradient as you suggested one from left-right and other from-right to left and vibrating animation.
![25state1](https://cloud.githubusercontent.com/assets/6409210/10018704/8a1c37ac-6141-11e5-8943-f51bdb8d65f9.png)
![25dotsstate25](https://cloud.githubusercontent.com/assets/6409210/10018804/325d05fe-6142-11e5-9a3c-de66cc165e24.png)

I suggested 25 dots as the maximum number of dots. But after experimenting with small device screen looks like 25 are to many dots and i had to find number of dots that can just fit the width of the card(as you know in small screen card takes 100% width of the device).And it looks like 18 fits perfectly.
![fadeleft](https://cloud.githubusercontent.com/assets/6409210/10018783/1a0c8cc2-6142-11e5-9d53-3b17e9fa187a.png)
Above image shows when you are at state 23 and the left-most dot is fading out to the left(it also shake in the interverval of 1s) to show that there are dots hidded to the left.

This shows fading out to the right 
![faderight2](https://cloud.githubusercontent.com/assets/6409210/10018950/0f031d4a-6143-11e5-93e5-bcbecdbf03c6.png)

One thing i have avoided is to add fading effects to the current dot . The clickable arrow(active arrow) can assist to show there are dots to the left/right. 
Thanks!

amitdeutsch commented:
Hi @MAKOSCAFEE,

Very nice work on this! One thing I would recommend: currently, the dots seem to fade to white, i.e. rgb(255,255,255), but instead they should fade to being completely transparent, i.e. rgba(255,255,255,0). Do you want to give that a try? 

Thanks,
Amit

MAKOSCAFEE commented:
Hi @amitdeutsch 
Sure! I changed to transparent and it easily noticeable than before.
![screenshot from 2015-09-22 18 48 08](https://cloud.githubusercontent.com/assets/6409210/10023452/eae959ca-615a-11e5-8f26-5108346a9c82.png)

Thanks!.

amitdeutsch commented:
Ah, very close. The gradients just need to be flipped; the dots should fade
*towards* the arrows, not *away* from the arrows.

On Tue, Sep 22, 2015 at 8:52 AM Barnabas Makonda <notifications@github.com>
wrote:

> Hi @amitdeutsch <https://github.com/amitdeutsch>
> Sure! I changed to transparent and it easily noticeable than before.
> [image: screenshot from 2015-09-22 18 48 08]
> <https://cloud.githubusercontent.com/assets/6409210/10023452/eae959ca-615a-11e5-8f26-5108346a9c82.png>
>
> Thanks!.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/oppia/oppia/pull/977#issuecomment-142331318>.
>

MAKOSCAFEE commented:
Thsnks @amitdeutsch 
![leftleft](https://cloud.githubusercontent.com/assets/6409210/10023740/23ffb9d8-615c-11e5-9444-2fea3b5ef195.png)
![rightgit](https://cloud.githubusercontent.com/assets/6409210/10023742/243dd768-615c-11e5-8804-95c45d3ad5a3.png)

Now can be seen with easy in the right directions.
Thanks!

amitdeutsch commented:
There we go! And, one more thing– sorry to add to the complexity– but the
fading should only occur when there are more cards in the history than
there are maximum dots; the 'fading' should denote that there are more
dots, not that you're next to an arrow. Does that make sense?

On Tue, Sep 22, 2015 at 9:02 AM Barnabas Makonda <notifications@github.com>
wrote:

> Thsnks @amitdeutsch <https://github.com/amitdeutsch>
> [image: leftleft]
> <https://cloud.githubusercontent.com/assets/6409210/10023740/23ffb9d8-615c-11e5-9444-2fea3b5ef195.png>
> [image: rightgit]
> <https://cloud.githubusercontent.com/assets/6409210/10023742/243dd768-615c-11e5-8804-95c45d3ad5a3.png>
>
> Now can be seen with easy in the right directions.
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/oppia/oppia/pull/977#issuecomment-142333778>.
>

MAKOSCAFEE commented:
Owwhh i have used two dots as maximum just to show how the new fading occur... I was lazy to create more than 18 states exploration. So i changed the maximum to 2. But practically fading will occur when you reach 19th state. 
see the screenshots  below
![18state](https://cloud.githubusercontent.com/assets/6409210/10027732/3c810a1a-6171-11e5-803d-71f827110b0b.png)
above is when you are in 18th state. you can see no fading anywhere
![19state](https://cloud.githubusercontent.com/assets/6409210/10027734/42646e40-6171-11e5-83ea-527289e00b66.png)
19th state and fading occurs to the left-most.
![20state](https://cloud.githubusercontent.com/assets/6409210/10027793/8ffe2682-6171-11e5-87e0-44c17957d5ea.png)
above there are 19 state but if i go to state 2 then fading occurs to the right.

Sorry for confusion.

seanlip commented on extensions/skins/conversation_v1/Conversation.js:
```diff
@@ -646,3 +646,41 @@ oppia.directive('progressDots', [function() {
+oppia.filter('sliceDots', function() {
+
+  return function(input, currentDotIndex) {
+    var _maximumNumberOfDotsShown = 18;
```
This is a constant, so use caps: MAX_DOTS

MAKOSCAFEE replied:
Done!

seanlip commented on extensions/skins/conversation_v1/Conversation.js:
```diff
@@ -646,3 +646,41 @@ oppia.directive('progressDots', [function() {
     }]
   };
 }]);
+oppia.filter('sliceDots', function() {
```
I don't think this should be a separate filter. Can the behaviour be folded into the progressDots directive?

The code in the filter also looks fairly complex. It might be cleaner to keep track of just leftmostVisibleDotIndex and rightmostVisibleDotIndex in the JS. Then, in the HTML, show a dot only if it's within this range, and similarly use these variables as the basis for the fadeout treatment, arrow disabling, etc.

MAKOSCAFEE replied:
I have merged the behavior into progressDots directive.
This also made the logic implementation to be easy.
Done!.

seanlip commented on extensions/skins/conversation_v1/player.html:
```diff
@@ -202,6 +202,21 @@ <h4 class="conversation-skin-responses-dropdown-container" ng-click="toggleShowP
+      50% { -webkit-transform: scale(0.9); }
+      0% { -webkit-transform: scale(1); }
+    }
+    @-moz-keyframes oppia-animate-bounce-dot  {
```
[This page](http://www.w3schools.com/cssref/css3_pr_animation-keyframes.asp) seems to suggest only -webkit is needed.

MAKOSCAFEE replied:
Done

seanlip commented on extensions/skins/conversation_v1/player.html:
```diff
@@ -230,6 +245,26 @@ <h4 class="conversation-skin-responses-dropdown-container" ng-click="toggleShowP
+      background: -o-linear-gradient(left,#ccc, rgba(255,255,255,0));
+      background: -moz-linear-gradient(left,#ccc, rgba(255,255,255,0));
+      background: linear-gradient(left,#ccc, rgba(255,255,255,0));
+      animation:oppia-animate-bounce-dot 1s linear 0s infinite ;
```
nit: add space after colon, delete space before semicolon.

MAKOSCAFEE replied:
Done

seanlip commented on extensions/skins/conversation_v1/player.html:
```diff
@@ -230,6 +245,26 @@ <h4 class="conversation-skin-responses-dropdown-container" ng-click="toggleShowP
+      background: -webkit-linear-gradient(right,#ccc,rgba(255,255,255,0));
+      background: -o-linear-gradient(right,#ccc, rgba(255,255,255,0));
+      background: -moz-linear-gradient(right,#ccc, rgba(255,255,255,0));
+      background: linear-gradient(right,#ccc, rgba(255,255,255,0));
```
forgot 'animation' rule?

MAKOSCAFEE replied:
Sure!..Its done!

seanlip commented on extensions/skins/conversation_v1/player.html:
```diff
@@ -258,11 +297,13 @@ <h4 class="conversation-skin-responses-dropdown-container" ng-click="toggleShowP
     </li>
-    <li ng-repeat="dot in dots track by $index" class="oppia-animate-progress-dot">
+    <li ng-repeat="dot in dots | sliceDots:currentDotIndex  track by $index"
+        ng-class="dot.isVisible ? 'oppia-animate-progress-dot': 'oppia-animate-progress-dot-hidden'">
```
I realize I suggested ng-class -- but actually Angular has a built-in directive, ng-show, for this sort of thing. It might actually be nicer to use ng-show in here based on leftmostVisibleDotIndex and rightmostVisibleDotIndex as described above.

MAKOSCAFEE replied:
this worked like charm.
Done

seanlip commented on extensions/skins/conversation_v1/static/conversation.css:
```diff
@@ -132,6 +132,8 @@ md-card.conversation-skin-tutor-card {
 .conversation-skin-tutor-card-middle-section .conversation-skin-oppia-avatar {
   left: -56px;
   top: -7px;
+  /* This is needed to prevent overlapping of avatars*/
```
nit: add full stop at end, add space before */

MAKOSCAFEE replied:
Done

seanlip commented on extensions/skins/conversation_v1/static/conversation.css:
```diff
@@ -142,6 +144,8 @@ md-card.conversation-skin-tutor-card {
 .conversation-skin-oppia-feedback-container {
   margin-bottom: 30px;
   position: relative;
+  /* This is needed to align content in-line with avatars*/
```
Ditto.

MAKOSCAFEE replied:
Done

seanlip commented on extensions/skins/conversation_v1/Conversation.js:
```diff
@@ -643,6 +646,25 @@ oppia.directive('progressDots', [function() {
+
+      /* This is needed to watch any change for both number of dots and currentDotIndex. */
+      $scope.$watchGroup(['dots', 'currentDotIndex'], function(oldValue, newValue) {
+        if(newValue !== oldValue) {
```
nit: here and below, add single space after 'if'

MAKOSCAFEE replied:
Done

seanlip commented on extensions/skins/conversation_v1/Conversation.js:
```diff
@@ -643,6 +646,25 @@ oppia.directive('progressDots', [function() {
+            }
+          } else{
+            $scope.leftmostVisibleDotIndex = 0;
+            $scope.rightmostVisibleDotIndex = $scope.MAX_DOTS;
```
Not $scope.dots.length - 1?

MAKOSCAFEE replied:
actually it is ```$scope.dots.length```  because in html i have used ```< $scope.rightmostVisibleDotIndex```

MAKOSCAFEE replied:
Done

seanlip commented on extensions/skins/conversation_v1/Conversation.js:
```diff
@@ -643,6 +646,25 @@ oppia.directive('progressDots', [function() {
+        if(newValue !== oldValue) {
+          if($scope.dots.length > $scope.MAX_DOTS) {
+            if($scope.currentDotIndex < $scope.dots.length - $scope.MAX_DOTS) {
+              $scope.leftmostVisibleDotIndex = $scope.currentDotIndex;
```
Actually I'm not sure this is right. Suppose there are 20 cards (0-indexed from 0 to 19). The user goes to card 2 then clicks the left arrow then the left arrow again (so, now on card 0). The user now clicks card 8 (say). The dots showing should still be 0 to 17.

How about instead having changeActiveDot(), decrementCurrentDotIndex(), incrementCurrentDotIndex() make the appropriate adjustments to leftmostVisibleDotIndex and rightmostVisibleDotIndex?

MAKOSCAFEE replied:
Done!.
Eg. i have created 22 state exploration
![last](https://cloud.githubusercontent.com/assets/6409210/10065970/ffb2668a-6295-11e5-9f8d-f32f41f9e80d.png)

but if you go to index 0(state 1). 
![0twent2](https://cloud.githubusercontent.com/assets/6409210/10065801/ed93fbc8-6293-11e5-81ec-4bf83c2ecb0b.png)
the range will stay the same (0-17 for this example case) untill ```currentDotPosition```  is at index that is greater than ```HALF_DOTS_RANGE```. snapshot below shows  11 state and you can see index zero does not fade and left arrow is active to navigate left as ussual 
![eleven](https://cloud.githubusercontent.com/assets/6409210/10065940/9a78ddee-6295-11e5-80bb-a84426614d34.png)

and if  ```currentDotPosition``` is at the position greater than ```HALF_DOTS_RANGE``` to the left and right fading effects will in both sides and ``currentDotPosition``` will be at the middle.
![3to20](https://cloud.githubusercontent.com/assets/6409210/10065883/f23c9364-6294-11e5-80da-6e41daedd099.png)
in the above means there are 3 dots hidden to the left(0-2) and 2 to the right(21,22).  I guess this was desired results?, if not I will be more than happy to make it better.

seanlip commented on extensions/skins/conversation_v1/player.html:
```diff
@@ -258,11 +276,16 @@ <h4 class="conversation-skin-responses-dropdown-container" ng-click="toggleShowP
-    <li ng-repeat="dot in dots track by $index" class="oppia-animate-progress-dot">
+    <li ng-repeat="dot in dots track by $index"
+        class="oppia-animate-progress-dot"
+        ng-show="($index >= leftmostVisibleDotIndex && $index < rightmostVisibleDotIndex)" >
```
No need brackets. Also I suggest that rightmostVisibleDotIndex be an inclusive bound (since, given its name, you'd expect it to be one of the dots showing on the page).

seanlip commented on extensions/skins/conversation_v1/player.html:
```diff
@@ -258,11 +276,16 @@ <h4 class="conversation-skin-responses-dropdown-container" ng-click="toggleShowP
       </span>
       <span class="oppia-progress-dot oppia-progress-dot-inactive"
+            ng-class="($index === leftmostVisibleDotIndex && $index !== 0) ? 'oppia-progress-dot-gradient-left':
+                      ($index === rightmostVisibleDotIndex-1 && $index !== MAX_DOTS-1) ?
```
Ditto re off-by-one.

seanlip commented on extensions/skins/conversation_v1/player.html:
```diff
@@ -258,11 +276,16 @@ <h4 class="conversation-skin-responses-dropdown-container" ng-click="toggleShowP
       <span class="oppia-progress-dot oppia-progress-dot-inactive"
+            ng-class="($index === leftmostVisibleDotIndex && $index !== 0) ? 'oppia-progress-dot-gradient-left':
+                      ($index === rightmostVisibleDotIndex-1 && $index !== MAX_DOTS-1) ?
+                       'oppia-progress-dot-gradient-right' : null "
```
nit: no space before "

Should null be empty string?

MAKOSCAFEE commented:
@seanlip  This actually update the range only when user answers the question or using increment and decrement functions when appropiate. see if it does what you have in mind and i can clean the code and add tootip

seanlip commented on extensions/skins/conversation_v1/Conversation.js:
```diff
@@ -621,6 +622,12 @@ oppia.directive('progressDots', [function() {
           $scope.currentDotIndex = $scope.dots.length - 1;
+          $scope.rightmostVisibleDotIndex = $scope.dots.length - 1;
+          if ($scope.dots.length > $scope.MAX_DOTS) {
+            $scope.leftmostVisibleDotIndex = $scope.rightmostVisibleDotIndex - $scope.MAX_DOTS;
```
off by one

MAKOSCAFEE replied:
Done

seanlip commented on extensions/skins/conversation_v1/Conversation.js:
```diff
@@ -634,15 +641,24 @@ oppia.directive('progressDots', [function() {
 
       $scope.incrementCurrentDotIndex = function() {
         if ($scope.currentDotIndex < $scope.dots.length - 1) {
+          if($scope.currentDotIndex === $scope.rightmostVisibleDotIndex){
```
nit: add space after 'if'

seanlip commented on extensions/skins/conversation_v1/player.html:
```diff
@@ -258,11 +276,16 @@ <h4 class="conversation-skin-responses-dropdown-container" ng-click="toggleShowP
       </span>
       <span class="oppia-progress-dot oppia-progress-dot-inactive"
+            ng-class="$index === leftmostVisibleDotIndex && $index !== 0 ? 'oppia-progress-dot-gradient-left':
+                      $index === rightmostVisibleDotIndex && $index !== dots.length-1 ?
```
nit: add single spaces around '-'

MAKOSCAFEE replied:
Done
